src/lib/Inference.hs

```diff
@@ -2629,6 +2654,53 @@ synthTopBlock block = do
   (liftExcept =<<) $ liftDictSynthTraverserM $ traverseGenericE block
 {-# SCC synthTopBlock #-}
 
+-- Given a simplified dict (an Atom of type `DictTy _` in the
+-- post-simplification IR), and a requested, more general, dict type, generalize
+-- the dict to match the more general type. This is only possible because we
+-- require that instances are polymorphic in data-role parameters.
+generalizeDict :: (EnvReader m) => Type n -> Dict n-> m n (Dict n)
+generalizeDict ty dict = do
+  result <- liftSolverM $ solveLocal $ generalizeDictAndUnify (sink ty) (sink dict)
+  case result of
+    Failure e -> error $ pprint e
+    Success ans -> return ans
+
+generalizeDictAndUnify :: EmitsInf n => Type n -> Dict n -> SolverM n (Dict n)
+generalizeDictAndUnify ty dict = do
+  dict' <- generalizeDictRec dict
+  dictTy <- getType dict'
+  unify ty dictTy
+  zonk dict'
+
+generalizeDictRec :: EmitsInf n => Dict n -> SolverM n (Dict n)
+generalizeDictRec dict = do
+  -- TODO: we should be able to avoid the normalization here . We only need it
+  -- because we sometimes end up with superclass projections. But they shouldn't
+  -- really be allowed to occur in the post-simplification IR.
+  DictCon dict' <- cheapNormalize dict
+  DictCon <$> case dict' of
+    InstanceDict instanceName args -> do
+      InstanceDef _ bs _ _ <- lookupInstanceDef instanceName
+      args' <- generalizeInstanceArgs bs args
+      return $ InstanceDict instanceName args'
+    IxFin _ -> IxFin <$> Var <$> freshInferenceName NatTy
+    ExplicitMethods _ _ _ -> notSimplifiedDict
+    InstantiatedGiven _ _ -> notSimplifiedDict
+    SuperclassProj _ _    -> notSimplifiedDict
+    where notSimplifiedDict = error $ "Not a simplified dict: " ++ pprint dict
+
+generalizeInstanceArgs :: EmitsInf n => Nest RolePiBinder n l -> [Atom n] -> SolverM n [Atom n]
+generalizeInstanceArgs Empty [] = return []
+generalizeInstanceArgs (Nest (RolePiBinder b ty _ role) bs) (arg:args) = do
+  arg' <- case role of
+    TypeParam -> Var <$> freshInferenceName TyKind
+    DictParam -> generalizeDictAndUnify ty arg
+    DataParam -> Var <$> freshInferenceName ty
```

Comment:
I am a little surprised by how uniform the treatment of type and data parameters ended up, because they should act quite differently here. It is definitely ok to have instances that are not completely polymorphic over the type parameters. Or is it perhaps an instance of what I mentioned before, which is that we could try to detect how polymorphic a derivation was by looking at which terms are applied as instance args? I'd like to understand this point a bit better, because it does make me uneasy.

Reply:
Ok I think I understand why this code might be looking like this, and what's the source of potential issues. One option would be to leave the type unmodified, but this could lead to potential unification issues if you've generalized their data components. Ideally, we'd traverse the type here and only leave the inference vars in place of the data components, since actually we do care about the non-data backbone of this type to be equal (and any inequality should lead to errors!). So it's really an unsound shortcut that might work, but we shouldn't be doing this IMO. Those functions should really live in `Generalize.hs` and then you'd be able to share the type traversal between them.

Reply:
Yes, I agree that would be cleaner and more obviously sound. I'll do that.